### PR TITLE
[client] Asyncio Support

### LINKED ISSUE #13747: [ray_client] Make ClientObjectRefs awaitable
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
```python
import asyncio

import ray

ray.util.connect("localhost:61234")

@ray.remote
def f():
    return "Hello!"

async def main():
    print(ray.get(f.remote())) # Works.
    print(await f.remote()) # Fails.

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

```
(base) code/ray % python test.py 
Hello!
Traceback (most recent call last):
  File "test.py", line 17, in <module>
    loop.run_until_complete(main())
  File "/Users/eoakes/anaconda3/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "test.py", line 14, in main
    print(await f.remote())
TypeError: object ClientObjectRef can't be used in 'await' expression
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
* Adds asyncio support to the ClientObjectRef in the form of `__await__` (making it 'awaitable') and `as_future()` 
* It also adds the private helper method `_on_completed` to bring it to parity with `ObjectRef.pxi`

## Related issue number
Closes #13747

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests <-- NOTE: This just reuses `test_asyncio.py`

### CODE REVIEW

File: python/ray/util/client/dataclient.py [hunk 1]
```diff
@@ -66,9 +70,19 @@ def _data_main(self) -> None:
                     # This is not being waited for.
                     logger.debug(f"Got unawaited response {response}")
                     continue
-                with self.cv:
-                    self.ready_data[response.req_id] = response
-                    self.cv.notify_all()
+                if response.req_id in self.asyncio_waiting_data:
+                    fut = self.asyncio_waiting_data[response.req_id]
+                    del self.asyncio_waiting_data[response.req_id]
```
Comment (simon-mo) [inline python/ray/util/client/dataclient.py:75]: fut = self.asyncio_waiting_data.pop(response.req_id)

**Comment by simon-mo** [comment]
Per offline discussion:
```
class ClientObjectRef(ClientBaseRef):
    def __await__(self):
        loop = asyncio.get_event_loop()
        fut = loop.create_future()

        def set_value(f):
            from ray.util.client.client_pickler import loads_from_server
            loop.call_soon_threadsafe(
                lambda : fut.set_result(loads_from_server(f.result().get.data))
            )
            

        ray._asyncio_get(self, set_value)
        return fut.__await__()

    def _on_completed(self, cb):
        ray._asyncio_get(self, cb)

-----------
class DataClient:
    if response.req_id in self.asyncio_waiting_data:
        cb_list = self.asyncio_waiting_data.pop(response.req_id)
        for cb in cb_list:
            cb(deserialized_data)


```

File: python/ray/util/client/common.py [hunk 2]
```diff
@@ -60,7 +62,41 @@ def __del__(self):
...
 class ClientObjectRef(ClientBaseRef):
-    pass
+    def __await__(self):
+        return self.as_future().__await__()
+
+    def as_future(self):
+        loop = asyncio.get_event_loop()
+        fut = loop.create_future()
+
+        def set_value(resp):
```
Comment (simon-mo) [inline python/ray/util/client/common.py:72]: Can we just use _on_completed with a simpler callback? So the deserialization code isn't duplicated. Something like
```
def set_result(obj):
    def inner():
        if isinstance(obj, Exception):
            fut.set_exception(obj)
        ...
     loop.call_soon_threadsafe(inner)
    
```

File: python/ray/util/client/common.py [hunk 3]
```diff
@@ -60,7 +62,41 @@ def __del__(self):
...
+            from ray.util.client.client_pickler import loads_from_server
+
+            def inner_set_value():
+                obj = resp.get
```
Comment (simon-mo) [inline python/ray/util/client/common.py:76]: .get()?

Comment (ijrsvt) [reply python/ray/util/client/common.py:76]: Nope, it is a property

File: python/ray/util/client/common.py (hunk 2 above)
Comment (ijrsvt) [reply python/ray/util/client/common.py:72]: Ahh yep, I think so!!

**Review by simon-mo — APPROVED** [review]

File: python/ray/util/client/dataclient.py [hunk 4]
```diff
@@ -35,6 +34,11 @@ def __init__(self, channel: "grpc._channel.Channel", client_id: str,
         self.ready_data: Dict[int, Any] = {}
         self.cv = threading.Condition()
         self.lock = threading.RLock()
+
+        # NOTE: Dictionary insertion is guaranteed to complete before lookup
+        # and/or removal because of synchronization via the request_queue.
+        self.asyncio_waiting_data: Dict[int, Callable[
+            [ray_client_pb2.DataResponse], None]] = {}
```
Comment (ijrsvt) [inline python/ray/util/client/dataclient.py:41]: @simon-mo I couldn't find official Pydocs on Dictionaries + Thread Safety, but IIRC python is okay with multi-threaded access as long as operations on the same key do not have race conditions.

Comment (simon-mo) [reply python/ray/util/client/dataclient.py:41]: oh.. it's definitely protected by GIL. dict.pop() is threadsafe i believe

**Review by AmeerHajAli — COMMENTED** [review]
Thanks for this PR @ijrsvt this was simpler than I expected!

I am concerned about putting this under ClientObjectRef class and not the top-level class (ClientBaseRef). For Example, are ClientActorRefs awaitable?
Can we add tests for actors and tasks? 
Did we check if the failing test in Serve works now? Maybe we can add it to the CI in this PR?

**Comment by ijrsvt** [comment]
@AmeerHajAli 

> I am concerned about putting this under ClientObjectRef class and not the top-level class (ClientBaseRef). For Example, are ClientActorRefs awaitable?

No, I put this in **ClientObjectRef** because Actor Refs are not awaitable. This also precisely mirrors the case for an ObjectRef [(see object_ref.pxi)](https://github.com/ray-project/ray/blob/5f0be94989030f0a796c49ec14b5a96878a5ba06/python/ray/includes/object_ref.pxi#L70).

> Can we add tests for actors and tasks?

Instead of writing new tests, the `test_asyncio.py` tests run for client as well, meaning that we have the same coverage as non-client mode. 

> Did we check if the failing test in Serve works now? Maybe we can add it to the CI in this PR?
I can add a separate test for serve.

**Review by AmeerHajAli — APPROVED** [review]

**Comment by AmeerHajAli** [comment]
Awesome!

**Comment by AmeerHajAli** [comment]
@ijrsvt by separate test, do you mean the test that failed previously when we did our discovery of unknowns?

**Comment by ijrsvt** [comment]
@AmeerHajAli @simon-mo this PR is only 50% async...The server side does not have any `async` behavior, meaning that when the client calls `await x` , the server will block until `x` is ready. This means that the client can do non-ray work, but any requests requiring that connection will be blocked until `x` is completed.

**Comment by AmeerHajAli** [comment]
> @AmeerHajAli @simon-mo this PR is only 50% async...The server side does not have any `async` behavior, meaning that when the client calls `await x` , the server will block until `x` is ready. This means that the client can do non-ray work, but any requests requiring that connection will be blocked until `x` is completed.

@ijrsvt , what do you suggest to resolve this issue and avoid server blocking behavior?

**Comment by ijrsvt** [comment]
@AmeerHajAli I spent a lot of Friday trying to prototype a few ideas. I'll meet with @simon-mo later today and will put together a quick list of alternatives!

**Comment by AmeerHajAli** [comment]
@ijrsvt , is this ready for review again?

**Comment by ijrsvt** [comment]
@AmeerHajAli @simon-mo yes, this is ready for re-review!

File: python/ray/util/client/dataclient.py [hunk 5]
```diff
@@ -66,9 +70,16 @@ def _data_main(self) -> None:
...
-                with self.cv:
-                    self.ready_data[response.req_id] = response
-                    self.cv.notify_all()
+                if response.req_id in self.asyncio_waiting_data:
+                    cb = self.asyncio_waiting_data.pop(response.req_id)
+                    try:
+                        cb(response)
+                    except Exception:
+                        logger.exception("Callback error:")
+                else:
```
Comment (simon-mo) [inline python/ray/util/client/dataclient.py:79]: a regular ref can have both `_on_completed` and `ray.get`. I don't think we need the else here.

File: python/ray/util/client/server/server.py [hunk 6]
```diff
@@ -252,6 +254,32 @@ def Terminate(self, req, context=None):
     def GetObject(self, request, context=None):
         return self._get_object(request, "", context)
 
+    def _async_get_object(self, request, client_id: str, req_id: int, context):
+        if request.id not in self.object_refs[client_id]:
+            return ray_client_pb2.GetResponse(valid=False)
+        try:
+            assert request.asynchronous
+            assert context is not None
+            objectref = self.object_refs[client_id][request.id]
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:263]: naming: either ref or object_ref

File: python/ray/util/client/server/server.py [hunk 7]
```diff
@@ -252,6 +254,32 @@ def Terminate(self, req, context=None):
...
+            logger.debug("async get: %s" % objectref)
+            with disable_client_hook():
+                loop = asyncio.get_event_loop()
+
+                def f(result):
+                    serialized = dumps_from_server(result, client_id, self)
+                    resp = ray_client_pb2.DataResponse(
+                        get=ray_client_pb2.GetResponse(
+                            valid=True, data=serialized))
+                    resp.req_id = req_id
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:273]: can this fail/throw?

File: python/ray/util/client/server/server.py [hunk 8]
```diff
@@ -556,7 +584,11 @@ def default_connect_handler(job_config: JobConfig = None):
         logs_servicer=logs_servicer,
         grpc_server=server,
     )
-    server.start()
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(server.start())
+    t = threading.Thread(
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:589]: might need to make it daemon=True so it won't block the client_server from shutdown

File: python/ray/util/client/server/server.py [hunk 9]
```diff
@@ -556,7 +584,11 @@ def default_connect_handler(job_config: JobConfig = None):
         logs_servicer=logs_servicer,
         grpc_server=server,
     )
-    server.start()
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(server.start())
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:588]: oof mixing event loop between threads can lead to hard to debug issues.

File: python/ray/util/client/server/server.py [hunk 10]
```diff
@@ -252,6 +254,32 @@ def Terminate(self, req, context=None):
...
+                loop = asyncio.get_event_loop()
+
+                def f(result):
+                    serialized = dumps_from_server(result, client_id, self)
+                    resp = ray_client_pb2.DataResponse(
+                        get=ray_client_pb2.GetResponse(
+                            valid=True, data=serialized))
+                    resp.req_id = req_id
+
+                    loop.create_task(context.write(resp))
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:275]: this function will be called from a core worker thread, so you probably need `call_soon_threadsafe`

File: python/ray/util/client/server/server.py [hunk 11]
```diff
@@ -252,6 +254,32 @@ def Terminate(self, req, context=None):
...
+            return ray_client_pb2.GetResponse(valid=False)
+        try:
+            assert request.asynchronous
+            assert context is not None
+            objectref = self.object_refs[client_id][request.id]
+            logger.debug("async get: %s" % objectref)
+            with disable_client_hook():
+                loop = asyncio.get_event_loop()
+
+                def f(result):
```
Comment (simon-mo) [inline python/ray/util/client/server/server.py:268]: naming

File: python/ray/util/client/server/server.py (hunk 9 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:588]: resolved offline. But will use separate asyncio event loop.

File: python/ray/util/client/server/server.py (hunk 8 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:589]: Done!

File: python/ray/util/client/server/server.py (hunk 11 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:268]: +1, will switch to `send_get_response`

File: python/ray/util/client/server/server.py (hunk 10 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:275]: Will switch to call_soon_threadsafe!

File: python/ray/util/client/server/server.py (hunk 7 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:273]: In the normal use case it does not appear to be in a try except. That being said, I'll nest the `dumps_from_server` inside a `try except` block to ensure we don't have anything blowing up in the CoreWorker.

File: python/ray/util/client/dataclient.py (hunk 5 above)
Comment (ijrsvt) [reply python/ray/util/client/dataclient.py:79]: Those would be under different request IDs. `ready_data` is used for signaling to `blocking send`.

File: python/ray/tests/test_client.py [hunk 12]
```diff
@@ -482,7 +482,7 @@ def run_client():
     time.sleep(3)
     server = ray_client_server.serve("localhost:50051")
     thread.join()
-    server.stop(0)
+    server.shutdown(0)
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client.py:485]: why?

File: python/ray/util/client/common.py [hunk 13]
```diff
@@ -60,7 +62,48 @@ def __del__(self):
...
+
+    def _on_completed(self, py_callback: Callable[[Any], None]) -> None:
+        """Register a callback that will be called after Object is ready.
+        If the ObjectRef is already ready, the callback will be called soon.
+        The callback should take the result as the only argument. The result
+        can be an exception object in case of task error.
+        """
+        from ray.util.client.client_pickler import loads_from_server
+
+        def deserialize_obj(resp):
```
Comment (AmeerHajAli) [inline python/ray/util/client/common.py:95]: type annotation

File: python/ray/util/client/common.py [hunk 14]
```diff
@@ -60,7 +62,48 @@ def __del__(self):
 
 
 class ClientObjectRef(ClientBaseRef):
-    pass
+    def __await__(self):
+        return self.as_future().__await__()
+
+    def as_future(self):
```
Comment (AmeerHajAli) [inline python/ray/util/client/common.py:68]: type annotation

File: python/ray/util/client/common.py [hunk 15]
```diff
@@ -60,7 +62,48 @@ def __del__(self):
...
+        loop = asyncio.get_event_loop()
+        fut = loop.create_future()
+
+        def set_value(data):
```
Comment (AmeerHajAli) [inline python/ray/util/client/common.py:72]: type annotation

File: python/ray/util/client/common.py [hunk 16]
```diff
@@ -60,7 +62,48 @@ def __del__(self):
...
+            def inner_set_value():
```
Comment (AmeerHajAli) [inline python/ray/util/client/common.py:73]: type annotation

File: python/ray/util/client/ray_client_helpers.py [hunk 17]
```diff
@@ -40,4 +40,4 @@ def ray_connect_handler(job_config=None):
     finally:
         ray._inside_client_test = False
         ray.disconnect()
-        server.stop(0)
+        server.shutdown(0)
```
Comment (AmeerHajAli) [inline python/ray/util/client/ray_client_helpers.py:43]: y

File: python/ray/util/client/server/dataservicer.py [hunk 18]
```diff
@@ -48,8 +48,17 @@ def Datapath(self, request_iterator, context):
                 else:
                     assert accepted_connection
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:49]: what do you expect this assert to do?

File: python/ray/util/client/server/dataservicer.py [hunk 19]
```diff
@@ -25,7 +25,7 @@ def __init__(self, basic_service: "RayletServicer"):
         self.clients_lock = Lock()
         self.num_clients = 0  # guarded by self.clients_lock
 
-    def Datapath(self, request_iterator, context):
+    async def Datapath(self, request_iterator, context):
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:28]: can we split this function a bit for readability? there are plenty of nested ifs, and returns/breaks/continues in many places. I can't understand what this is doing without significant effort.

File: python/ray/util/client/server/server.py [hunk 20]
```diff
@@ -252,14 +254,46 @@ def Terminate(self, req, context=None):
     def GetObject(self, request, context=None):
         return self._get_object(request, "", context)
 
+    def _async_get_object(self, request, client_id: str, req_id: int, context):
+        if request.id not in self.object_refs[client_id]:
+            return ray_client_pb2.GetResponse(valid=False)
+        try:
+            assert request.asynchronous
+            assert context is not None
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/server.py:262]: do we need these?
What is context here?

File: python/ray/util/client/server/server.py [hunk 21]
```diff
@@ -252,14 +254,46 @@ def Terminate(self, req, context=None):
     def GetObject(self, request, context=None):
         return self._get_object(request, "", context)
 
+    def _async_get_object(self, request, client_id: str, req_id: int, context):
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/server.py:257]: type annotation

File: python/ray/util/client/server/server.py [hunk 22]
```diff
@@ -252,14 +254,46 @@ def Terminate(self, req, context=None):
...
+            return ray_client_pb2.GetResponse(valid=False)
+        try:
+            assert request.asynchronous
+            assert context is not None
+            object_ref = self.object_refs[client_id][request.id]
+            logger.debug("async get: %s" % object_ref)
+            with disable_client_hook():
+                loop = asyncio.get_event_loop()
+
+                def send_get_response(result):
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/server.py:268]: type annotation

File: python/ray/util/client/server/server.py (hunk 20 above)
Comment (AmeerHajAli) [reply python/ray/util/client/server/server.py:262]: since these server sides, they will crash the server and the client will just disconnect.

**Comment by AmeerHajAli** [comment]
Does the ray serve handle test pass now?

File: python/ray/util/client/server/dataservicer.py (hunk 18 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:49]: @AmeerHajAli I did not add this. But it just ensures `init` has been called before this Datapath is useed.

File: python/ray/util/client/server/server.py (hunk 20 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:262]: No, they are wrapped in a try-Except. They will just be returned to the client side as an error.

File: python/ray/util/client/server/dataservicer.py (hunk 19 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:28]: @AmeerHajAli I'd really prefer to push this into a later PR. This is essentially just a `large switch` statement over the possible request types.

If you want, I can move more of the request logic out, but I wanted to further nest it in this **main function** to make it clear that the `continue` logic _only ever happens_ for asynchronous requests.

**Comment by ijrsvt** [comment]
> Does the ray serve handle test pass now?

Yes, let me just add it to a client test.

**Comment by ijrsvt** [comment]
I am closing this PR because this approach has **2 major problems**
1. The server hangs if it is shutdown before all clients `disconnect`. The problem comes from the ThreadpoolExecutor that is passed to gRPC. Some of the threads created by this threadpool are stuck running and prevent python from exiting.
2. At a fundamental level, using `asyncio` for the `Dataclient` means that _different clients are on the same asyncio.EventLoop_. This means that the following will not work properly:
```
client1 = ray.connect(...)
ray.get(ray.remote(lambda: time.sleep(3000))
------------------------
client2 = ray.connect(...)
ray.get(ray.remote(lambda: 30)) # Will be blocked by the ray.get from client1
```

**Comment by AmeerHajAli** [comment]
What is the alternative ?

**Comment by ijrsvt** [comment]
@AmeerHajAli It is in #4242 But the plan is to use a synchronous server.

**Pull request closed without merging**